src/OVAL/probes/SEAP/sch_queue.c

```diff
@@ -0,0 +1,191 @@
+	}
+
+	desc->scheme_data = (void *) data;
+	return 0;
+}
+
+int sch_queue_openfd(SEAP_desc_t *desc, int fd, uint32_t flags)
+{
+	return 0;
+}
+
+int sch_queue_openfd2(SEAP_desc_t *desc, int ifd, int ofd, uint32_t flags)
+{
+	return 0;
+}
+
+ssize_t sch_queue_recv(SEAP_desc_t *desc, void *buf, size_t len, uint32_t flags)
+{
+	sch_queuedata_t *data = (sch_queuedata_t *)desc->scheme_data;
+	struct oscap_queue *queue;
+	pthread_mutex_t *mutex;
+	pthread_cond_t *cond;
+	int *cnt;
+	if (pthread_self() == data->parent_thread_id) {
+		queue = data->from_probe_queue;
+		mutex = &data->from_probe_mutex;
+		cond = &data->from_probe_cond;
+		cnt = &data->from_probe_cnt;
+	} else {
+		queue = data->to_probe_queue;
+		mutex = &data->to_probe_mutex;
+		cond = &data->to_probe_cond;
+		cnt = &data->to_probe_cnt;
+	}
+	pthread_mutex_lock(mutex);
+	while (*cnt == 0) {
+		pthread_cond_wait(cond, mutex);
+	}
+	char *item = oscap_queue_remove(queue);
+	(*cnt)--;
+	strncpy(buf, item, len);
+	ssize_t ret = strlen(item);
+	pthread_mutex_unlock(mutex);
+	return ret;
+}
+
+ssize_t sch_queue_send(SEAP_desc_t *desc, void *buf, size_t len, uint32_t flags)
+{
+	sch_queuedata_t *data = (sch_queuedata_t *)desc->scheme_data;
+
+	char *msg = malloc(len + 1);
+	strncpy(msg, buf, len);
+	struct oscap_queue *queue;
+	pthread_mutex_t *mutex;
+	pthread_cond_t *cond;
+	int *cnt;
+	if (pthread_self() == data->parent_thread_id) {
+		queue = data->to_probe_queue;
+		mutex = &data->to_probe_mutex;
+		cond = &data->to_probe_cond;
+		cnt = &data->to_probe_cnt;
+	} else {
+		queue = data->from_probe_queue;
+		mutex = &data->from_probe_mutex;
+		cond = &data->from_probe_cond;
+		cnt = &data->from_probe_cnt;
+	}
+	pthread_mutex_lock(mutex);
+	oscap_queue_add(queue, (void *) msg);
+	(*cnt)++;
+	pthread_cond_broadcast(cond);
+	pthread_mutex_unlock(mutex);
+	return len;
+}
+
+ssize_t sch_queue_sendsexp(SEAP_desc_t *desc, SEXP_t *sexp, uint32_t flags)
+{
+	sch_queuedata_t *data = (sch_queuedata_t *) desc->scheme_data;
+	strbuf_t *sb = strbuf_new(SEAP_STRBUF_MAX);
+	if (SEXP_sbprintf_t(sexp, sb) != 0) {
+		return -1;
+	}
+	char *msg = oscap_strdup(strbuf_cstr(sb));
+
+	ssize_t len = strbuf_length(sb);
+	msg[len] = '\0';
+	struct oscap_queue *queue;
+	pthread_mutex_t *mutex;
+	pthread_cond_t *cond;
+	int *cnt;
+	if (pthread_self() == data->parent_thread_id) {
+		queue = data->to_probe_queue;
+		mutex = &data->to_probe_mutex;
+		cond = &data->to_probe_cond;
+		cnt = &data->to_probe_cnt;
+	} else {
+		queue = data->from_probe_queue;
+		mutex = &data->from_probe_mutex;
+		cond = &data->from_probe_cond;
+		cnt = &data->from_probe_cnt;
+	}
+	pthread_mutex_lock(mutex);
+	oscap_queue_add(queue, (void *) msg);
+	(*cnt)++;
+	pthread_cond_broadcast(cond);
+	pthread_mutex_unlock(mutex);
+	strbuf_free(sb);
+	return len;
+}
+
+int sch_queue_close(SEAP_desc_t *desc, uint32_t flags)
+{
+	int ret = 0;
+	sch_queuedata_t *data = (sch_queuedata_t *) desc->scheme_data;
+	if (pthread_cancel(data->probe_thread_id) != 0) {
```

Comment:
I think using pthread_cancel is a wrong approach. It might be a pthread_join instead.